spring-integration-mongodb/src/main/java/org/springframework/integration/mongodb/outbound/MongoDbOutboundGateway.java

```diff
@@ -0,0 +1,201 @@
+import org.springframework.integration.handler.AbstractReplyProducingMessageHandler;
+import org.springframework.messaging.Message;
+import org.springframework.util.Assert;
+import org.springframework.util.CollectionUtils;
+
+import com.mongodb.DBObject;
+
+/**
+ * Makes outbound operations to query a MongoDb database using a {@link MongoOperations}
+ *
+ * @author Xavier Padró
+ * @since 5.0
+ */
+public class MongoDbOutboundGateway extends AbstractReplyProducingMessageHandler {
+
+	private volatile boolean initialized = false;
+
+	private static final String DEFAULT_COLLECTION_NAME = "data";
+
+	private volatile EvaluationContext evaluationContext;
+
+	private volatile String mongoDbQuery;
+
+	private volatile Expression mongoDbQueryExpression;
+
+	private volatile MongoOperations mongoTemplate;
+
+	private volatile MongoDbFactory mongoDbFactory;
+
+	private volatile MongoConverter mongoConverter;
+
+	private volatile boolean expectSingleResult = false;
+
+	private volatile Class<?> entityClass = DBObject.class;
+
+	private volatile String collectionName;
+
+	private volatile Expression collectionNameExpression = new LiteralExpression(DEFAULT_COLLECTION_NAME);
+
+	public MongoDbOutboundGateway(MongoDbFactory mongoDbFactory) {
+		Assert.notNull(mongoDbFactory, "mongoDbFactory must not be null.");
+		this.mongoDbFactory = mongoDbFactory;
+	}
+
+	public MongoDbOutboundGateway(MongoOperations mongoTemplate) {
+		Assert.notNull(mongoTemplate, "mongoTemplate must not be null.");
+		this.mongoTemplate = mongoTemplate;
+	}
+
+	public void setMongoDbQuery(String mongoDbQuery) {
+		Assert.hasText(mongoDbQuery, "mongoDbQuery must neither be null nor empty.");
+		this.mongoDbQuery = mongoDbQuery;
+	}
+
+	public void setMongoDbQueryExpression(Expression mongoDbQueryExpression) {
+		Assert.notNull(mongoDbQueryExpression, "mongoDbQueryExpression must neither be null nor empty.");
+		this.mongoDbQueryExpression = mongoDbQueryExpression;
+	}
+
+	public void setExpectSingleResult(boolean expectSingleResult) {
+		this.expectSingleResult = expectSingleResult;
+	}
+
+	public void setEntityClass(Class<?> entityClass) {
+		Assert.notNull(entityClass, "entityClass must not be null.");
+		this.entityClass = entityClass;
+	}
+
+	public void setCollectionName(String collectionName) {
+		Assert.notNull(collectionName, "collectionName cannot be null");
+		this.collectionName = collectionName;
+	}
+
+	public void setCollectionNameExpression(Expression collectionNameExpression) {
+		Assert.notNull(collectionNameExpression, "collectionNameExpression cannot be null");
+		this.collectionNameExpression = collectionNameExpression;
+	}
+
+	public void setMongoConverter(MongoConverter mongoConverter) {
+		Assert.notNull(mongoConverter, "mongoConverter cannot be null");
+		Assert.isNull(this.mongoTemplate,
+				"'mongoConverter' can not be set when instance was constructed with MongoTemplate");
+		this.mongoConverter = mongoConverter;
```

Comment:
OK. I got!
Sad, of course, but we can live with that.
Let's then leave this part as is - a setter for the `MongoConverter`!
We can reconsider refactor in the separate task. Luckily we are in the major version, so we can provide breaking changes. I mean move that `MongoConverter` to the ctor argument for all components and, therefore, change the common `MongoParserUtils.processCommonAttributes()` to populate ctor arg instead of setter's value.

Everything rest still looks like in a power. 
Correct me if you have some doubts.

Thanks